plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,534 @@
+const IDENTIFIES_TTL_SECONDS = 60 * 60 * 24
+const DELETE_EXPIRED_SALTS_INTERVAL_MS = 60 * 60 * 1000
+
+export async function cookielessServerHashStep(
+    runner: EventPipelineRunner,
+    event: PluginEvent
+): Promise<[PluginEvent | undefined]> {
+    // if events aren't using this mode, skip all processing
+    if (!event.properties?.[COOKIELESS_MODE_FLAG_PROPERTY]) {
+        return [event]
+    }
+
+    // if the team isn't allowed to use this mode, drop the event
+    const team = await runner.hub.teamManager.getTeamForEvent(event)
+    if (!team?.cookieless_server_hash_mode) {
+        // TODO log
+        return [undefined]
+    }
+    const teamTimeZone = team.timezone
+
+    const timestamp = event.timestamp ?? event.sent_at ?? event.now
+
+    // drop some events that aren't valid in this mode
+    if (!timestamp) {
+        // TODO log
+        return [undefined]
+    }
+    const { $session_id: sessionId, $device_id: deviceId } = event.properties
+    if (sessionId != null || deviceId != null) {
+        // TODO log
+        return [undefined]
+    }
+
+    if (event.event === '$alias') {
+        // TODO support these
+        return [undefined]
+    }
+
+    // if it's an identify event, it must have the sentinel distinct id
+    if (event.event === '$identify' && event.properties['$anon_distinct_id'] !== COOKIELESS_SENTINEL_VALUE) {
+        // TODO log
+        return [undefined]
+    }
+
+    const {
+        userAgent,
+        ip,
+        host,
+        timezone: eventTimeZone,
+        timestampMs,
+        teamId,
+        hashExtra,
+    } = getProperties(event, timestamp)
+    if (!userAgent || !ip || !host) {
+        // TODO log
+        return [undefined]
+    }
+
+    if (
+        team.cookieless_server_hash_mode === CookielessServerHashMode.Stateless ||
+        process.env.FORCE_STATELESS_COOKIELESS_MODE
+    ) {
+        if (event.event === '$identify' || event.distinct_id !== COOKIELESS_SENTINEL_VALUE) {
+            // identifies and post-identify events are not valid in the stateless mode, drop the event
+            return [undefined]
+        }
+
+        const hashValue = await doHash(runner.hub.db, {
+            timestampMs,
+            eventTimeZone,
+            teamTimeZone,
+            teamId,
+            ip,
+            host,
+            userAgent,
+            hashExtra,
+        })
+        const distinctId = hashToDistinctId(hashValue)
+        event.distinct_id = distinctId
+        event.properties['$device_id'] = distinctId
+        event.properties['$session_id'] = createStatelessSessionId(timestampMs, eventTimeZone, teamTimeZone, hashValue)
+
+        return [event]
+    } else {
+        // TRICKY: if a user were to log in and out, to avoid collisions, we would want a different hash value, so we store the set of identify event uuids for identifies
+        // ASSUMPTION: all events are processed in order, for this to happen we need them to be in the same kafka topic at this point
```

Comment:
Is there any case where this assumption would become false? Can we add the necessary ties to guarantee we won't break this assumption without fixing this?

Reply:
> Is there any case where this assumption would become false?

The happy case processes events strictly in order. The unhappy case does not: we have an overflow queue for badly behaved events, and these are not processed in order.

A lot of our stuff breaks in subtle ways in the overflow queue, e.g. person properties can be applied in the wrong order. This is a trade-off because the alternative would be simply dropping the events.

So yes - this assumption can break, but the vast majority of the time it holds, and when it does, many other things are broken too.